### ISSUE TITLE
CMSAttachMenu failing to register after 3.0.12

### ISSUE THREAD

**Issue by mrhanlon** [divio/django-cms#3936]
I noticed that after upgrading from 3.0.10 to 3.0.12 that a CMSAttachMenu was not registering on a page. It would periodically appear and disappear. In the advanced settings sometimes I would see zero, one, or two options for the same menu. I only have one CMSAttachMenu menu defined in a single app in my project. I attempted to delete and recreate the page to no avail. I noticed in the database, if there was only one option available in advanced settings, the page record showed `UserMenu` for `navigation_extenders`. If there were two options, it would be `UserMenu:190` or `UserMenu:193` depending which option I selected.

Ultimately I reverted to 3.0.10 and the issue went away.

My menu class:

```python
from menus.base import NavigationNode
from cms.menu_bases import CMSAttachMenu
from menus.menu_pool import menu_pool
from django.utils.translation import ugettext_lazy as _

class UserMenu(CMSAttachMenu):

    name = _('User Menu')

    def get_nodes(self, request):
        nodes = []

        menu_id = 1
        root_id = menu_id

        # root node
        greeting = 'Hello, {0}'.format(request.user.get_full_name() or request.user) if request.user.is_authenticated() else 'Log in'
        title = '<i class="fa fa-user hidden-xs hidden-md hidden-lg"></i><span class="hidden-sm">{0}</span>'
        n = NavigationNode(title.format(greeting), "/user/", menu_id)
        nodes.append(n)

        # anonymous menus
        menu_id += 1
        n = NavigationNode(_('Log in'), "/login/", menu_id, root_id, attr={'visible_for_authenticated':False})
        nodes.append(n)

        menu_id += 1
        n = NavigationNode(_('Register'), "/register/", menu_id, root_id, attr={'visible_for_authenticated':False})
        nodes.append(n)

        menu_id += 1
        n = NavigationNode(_('Help'), "/help/ticket/new/guest/", menu_id, root_id, attr={'visible_for_authenticated':False})
        nodes.append(n)

        # authenticated menus
        menu_id += 1
        dashboard_id = menu_id
        n = NavigationNode(_('Dashboard'), "/user/dashboard/", menu_id, root_id, attr={'visible_for_anonymous':False})
        nodes.append(n)

        menu_id += 1
        n = NavigationNode(_('Log out'), "/logout/", menu_id, root_id, attr={'visible_for_anonymous':False})
        nodes.append(n)

        # user section dashboard sub-menu
        menu_id += 1
        n = NavigationNode(_('Dashboard'), "/user/dashboard/", menu_id, dashboard_id, attr={'visible_for_anonymous':False})
        nodes.append(n)

        menu_id += 1
        n = NavigationNode(_('Projects'), "/user/projects/", menu_id, dashboard_id, attr={'visible_for_anonymous':False})
        nodes.append(n)

        menu_id += 1
        n = NavigationNode(_('Help Desk'), "/user/help/", menu_id, dashboard_id, attr={'visible_for_anonymous':False})
        nodes.append(n)

        menu_id += 1
        n = NavigationNode(_('Profile'), "/user/profile/", menu_id, dashboard_id, attr={'visible_for_anonymous':False})
        nodes.append(n)

        return nodes

menu_pool.register_menu(UserMenu)
```

**Comment by yakky** [comment]
@mrhanlon could you give #3950 a try?

**Comment by mrhanlon** [comment]
@yakky It looks like this fixed it for me. I had to edit the page to attach the menu again, but it is working as expected. Thanks!

**Comment by yakky** [comment]
@mrhanlon thanks for testing. Yes menu might need to be attached again as a spurious menu might have been attached before

**Comment by mbi** [comment]
FWIW I also stumbled unto this issue and can confirm https://github.com/divio/django-cms/pull/3950 fixes it.

**Comment by mkoistinen** [comment]
Thanks, @yakky for the fix and all those reporting!

**Comment by mbi** [comment]
I'm still seeing this in 3.0.13

**Comment by mbi** [comment]
@yakky ping?

I'm still seeing all `CMSAttachMenu` instances being suggested twice in the admin (and neither is working when attached to the page). This is happening both in 3.0.13 and 3.1.0. The latest version that doesn't seem to have this problem is 3.0.10

Am I doing something wrong?

![screen shot 2015-04-22 at 11 00 55](https://cloud.githubusercontent.com/assets/131808/7270797/ff86af12-e8de-11e4-95ce-827984606040.png)

**Comment by yakky** [comment]
Have you wiped out local cache and updated aldryn-apphooks-config too?

**Comment by mbi** [comment]
I had (and just have again) cleared the local cache.

I'm not using Apphook namespaces (my apps are attached only once), nor `aldryn-apphooks-config` or any other namespace configuration mechanism. Am I supposed to? The documentation doesn't seem to imply this is required.

**Comment by mbi** [comment]
@yakky ping? Should this be reopened?

**Comment by vstoykov** [comment]
I think yes, because we also experience this problem with 3.0.13.

First I need some clarification about how CMSAttachMenu should actually work if it is registered alone or in AppHook or in AppHook attached multiple times.

 - If it is simple CMSAttachMenu it will show in a dropdown for every page  (Which is normal)
 - If it is registered and attached to page by AppHook it will show n-times, where n is number of pages to which AppHook is attached (and because every page has draft and public version it shows actually n*2 after page publication).

I'm not sure if it needs to show at all in the dropdown when the menu is attached to page by AppHook, and to allow user to attach it twice by using one of it's instances.

Also when AppHook with menu is attached to a page then menu elements doesn't work as expected when using `show_sub_menu`. Sometimes elements are shown, some times not (switching between live and draft version of a page). With `show_menu` and showing the whole tree everything is ok.

**Comment by vstoykov** [comment]
What I've discovered is that when `MenuPool._mark_selected` is called this is before `NavExtender` modify the tree (before setting correct parent for every attached node and before removing all useless nodes).
The result is that selected node is removed from the tree (in most cases when visiting draft version of the page) and `show_submenu` can't display any nodes because there is no selected page and there are no ancestors.

@mbi Is your problem related only to `show_submenu` or to `show_menu` also?

**Comment by mbi** [comment]
@vstoykov I'm only using `show_menu` on this particular project.

**Comment by mkoistinen** [comment]
Please try with this branch: https://github.com/divio/django-cms/tree/issues/fix_breadcrumbs?files=1

**Comment by mbi** [comment]
@mkoistinen With that branch the menus seem to get rendered correctly (yay!) but I'm still getting duplicate entries in the "attached menu" dropdown in the page's advanced settings.

**Comment by vstoykov** [comment]
@mkoistinen for me it looks also ok.

About double rendered menus I think that the option for attaching menu when AppHook is attached need to be disabled in order people to not be confused.
Also `page.navigation_extenders` to be cleared because `CMSAttachedMenu.get_instances` will get correct pages through apphook and to prevent accidentally attaching the same menu again (which can double the menu items).

**Comment by dzonder** [comment]
I still see issues with app menus in 3.1.

Lets assume app is available as page at /shop/ and app menus contains available products.
When accessing urls which are not in app menus (i.e. /shop/addresses/, /shop/addresses/add) everything works fine and menu is visible.
But when accessing product (i.e. /shop/product/pslug1) app menus disappear.

3.0.10 worked fine but I needed Django>=1.7 support.

Is anyone still working on this?

**Comment by vstoykov** [comment]
@dzonder changes made by @mrhanlon are note merged yet and issue persist in current released versions. You can try the branch mentioned by him and tell if the problem is fixed there for you.

Until issue is resolved you can try a workaround by putting this in `menu.py` in some of your applications:

    from menus.menu_pool import menu_pool
    from menus.modifiers import Marker
    
    class FixNodesTree(Marker):
        """
        The purpose of this modifier is to workaround Django CMS bug
        """
        post_cut = False

        def modify(self, request, nodes, namespace, root_id, post_cut, breadcrumb):
            if not post_cut:
                menu_pool._mark_selected(request, nodes)
                return super(FixNodesTree, self).modify(request, nodes, namespace, root_id, post_cut, breadcrumb)
            return nodes

    menu_pool.register_modifier(FixNodesTree)

**Comment by Astone** [comment]
Yep, same here, in version 3.1.0. I see the double entries in the advances settings, and the menu's don't show up if the actual content is below a different page as the menu.

Strangely enough it does seem to work quite fine, unpublished, published and even when I log out. But as soon as I reload apache (or restart the debug server) the menu's disappear.

@vstoykov Thanks for your suggestion, but it doesn't seem to have any effect on the behavior of my menu's...

**Comment by mkoistinen** [comment]
Awesome feedback guys. To be clear, *most* of the issue has been solved in the above-mentioned PR for the 3.0.x branch. What remains is more of an ugly cosmetic issue with the Advanced Page Settings. I'll try to fix this asap.

Once these are committed to the Support 3.0.x branch, we'll port it up to 3.1 too.

**Comment by Astone** [comment]
Hi all, what is the status of this bug?

I'm developing a website that's supposed to go live on the 1st of June. It's almost finished except for the menu's.

They organize workshops, courses and trainings, so I made an application to manage these. They all share the same models, but they are divided in three categories. The urls of the courses are: /some-fixed-name/[category_slug]/[slug]. I created three menus that filter on the category: WorkshopMenu, CourseMenu, TrainingMenu. These should show up under the pages /workshops/, /courses/, /trainings/.

When I assign a menu to a page, it shows up properly and all works as expected, until I restart the webserver. Than they just disappear and the page admin shows that no menu is selected. It shows the same behaviour in development mode (manage.py runserver) and in live mode (service apache2 reload).

I don't really know how to proceed, all URLs are correct and unique and I don't seem to get any errors.

Any thoughts on this?

**Comment by jsma** [comment]
@mkoistinen Anything we can do to help get this to the finish line? Is the only thing left to fix the fact that the menus are shown multiple times in the admin?

**Comment by mkoistinen** [comment]
Sorry all, I've been particularly consumed with personal things lately (like moving to another country in less than 1 week). However, I am going to try to focus on this again starting.... now! =)

**Comment by jsma** [comment]
Thanks @mkoistinen ! About to move 3,000 miles myself, I feel your pain. I'll watch the PR https://github.com/divio/django-cms/pull/4042 and let us know there if you need any help testing etc.

**Comment by mkoistinen** [comment]
Anyone care to test the updated version of this branch: https://github.com/divio/django-cms/tree/issues/fix_breadcrumbs

**Comment by jsma** [comment]
Will be able to in a couple hours, thanks!

**Comment by yakky** [comment]
reopening this as it's not fixed yet

**Comment by yakky** [comment]
This should have been fixed in #4042
@jsma @Astone @vstoykov @mbi @mrhanlon @dzonder could you please give latest ``support/3.0.x`` branch a try?

**Comment by vstoykov** [comment]
I will test these changes tomorrow at work and will told you if my problems are gone.

**Comment by vstoykov** [comment]
With current `support/3.0.x` branch a can't see the problem. For me it looks ok.

**Comment by yakky** [comment]
Thanks @vstoykov !

**Comment by yakky** [comment]
Fixed in #4042
